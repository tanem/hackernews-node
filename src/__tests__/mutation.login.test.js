const path = require('path')
const bcrypt = require('bcryptjs')
const chance = require('chance').Chance(path.basename(__filename))
const request = require('supertest')

let httpServerUrl
let prisma

beforeAll(async () => {
  httpServerUrl = global.__HTTP_SERVER_URL__
  prisma = global.__PRISMA__
})

const name = chance.first()
const email = chance.email()
const password = chance.word()

beforeEach(async () => {
  await prisma.user.create({
    data: {
      name,
      email,
      password: await bcrypt.hash(password, 10),
    },
  })
})

test('successful login', async () => {
  const { body, status } = await request(httpServerUrl)
    .post('/')
    .send({
      query: `
      mutation {
        login(
          email: "${email}"
          password: "${password}"
        ) {
          token
          user {
            email
            links {
              url
              description
            }
          }
        }
      }
    `,
    })

  expect(status).toBe(200)
  expect(body).toMatchSnapshot({
    data: {
      login: {
        token: expect.any(String),
      },
    },
  })
})

test('user not found', async () => {
  const { body, status } = await request(httpServerUrl)
    .post('/')
    .send({
      query: `
      mutation {
        login(
          email: "uknown@example.com"
          password: "${password}"
        ) {
          token
          user {
            email
            links {
              url
              description
            }
          }
        }
      }
    `,
    })

  expect(status).toBe(200)
  expect(body).toMatchSnapshot()
})

test('invalid password', async () => {
  const { body, status } = await request(httpServerUrl)
    .post('/')
    .send({
      query: `
      mutation {
        login(
          email: "${email}"
          password: "invalid"
        ) {
          token
          user {
            email
            links {
              url
              description
            }
          }
        }
      }
    `,
    })

  expect(status).toBe(200)
  expect(body).toMatchSnapshot()
})
