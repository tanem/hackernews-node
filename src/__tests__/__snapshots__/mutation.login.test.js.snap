// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`invalid password 1`] = `
Object {
  "data": Object {
    "login": null,
  },
  "errors": Array [
    Object {
      "locations": Array [
        Object {
          "column": 9,
          "line": 3,
        },
      ],
      "message": "Invalid password",
      "path": Array [
        "login",
      ],
    },
  ],
}
`;

exports[`successful login 1`] = `
Object {
  "data": Object {
    "login": Object {
      "token": Any<String>,
      "user": Object {
        "email": "segerrid@oznel.nf",
        "links": Array [],
      },
    },
  },
}
`;

exports[`user not found 1`] = `
Object {
  "data": Object {
    "login": null,
  },
  "errors": Array [
    Object {
      "locations": Array [
        Object {
          "column": 9,
          "line": 3,
        },
      ],
      "message": "No such user found",
      "path": Array [
        "login",
      ],
    },
  ],
}
`;
